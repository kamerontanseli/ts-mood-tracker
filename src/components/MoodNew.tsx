import { last } from "ramda";
import * as React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { addMood, setNextMood } from "../store/actions/moodHistory";
import { IMood } from "../store/reducers/moodHistory";
import { IAction, IState } from "../store/types";

export interface IMoodNewProps {
  lastRecorded: IMood | null;
  onChange: (value: number) => void;
  onSubmit: (value: number) => void;
  value: number;
}

export const MoodNew: React.SFC<IMoodNewProps> = ({
  lastRecorded,
  onChange,
  onSubmit,
  value
}: IMoodNewProps) => {
  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => (
    e.preventDefault(), onSubmit(value)
  );
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void =>
    onChange(Number(e.currentTarget.value));
  return (
    <div>
      {!!lastRecorded && <p>Last recorded: {lastRecorded.value}</p>}
      <form data-test="form" onSubmit={handleSubmit}>
        <input
          type="range"
          onChange={handleChange}
          name="mood"
          required={true}
          min="1"
          max="10"
          data-test="moodBar"
        />
        <input type="submit" value="Save" />
      </form>
      <Link to="/">Cancel</Link>
    </div>
  );
};

export const mapStateToProps = (state: IState): object => ({
  lastRecorded: last(state.moodHistory.values),
  value: state.moodHistory.nextMoodValue
});

export const mapDispatchToProps = (
  dispatch: (action: IAction) => void
): object => ({
  onChange: (value: number) => dispatch(setNextMood(value)),
  onSubmit: (value: number) => dispatch(addMood({ date: Date.now(), value }))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MoodNew);
