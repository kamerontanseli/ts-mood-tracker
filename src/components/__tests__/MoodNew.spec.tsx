import * as React from "react";
import * as renderer from "react-test-renderer";
import { MoodNew, mapStateToProps, mapDispatchToProps } from "../MoodNew";

jest.mock("react-router-dom");

describe('mapStateToProps', () => {
  it('should return lastRecorded and value', () => {
    const state = {
      moodHistory: {
        values: [ {date: 1, value: 2} ],
        nextMoodValue: 5
      }
    }
    expect(mapStateToProps(state)).toEqual({
      lastRecorded: {date: 1, value: 2},
      value: 5
    })
  })
})

describe("MoodNew", () => {
  it("should render", () => {
    const output = renderer.create(
      <MoodNew
        onChange={() => void 0}
        onSubmit={() => void 0}
        value={20}
        lastRecorded={{ date: 1542226365908, value: 5 }}
      />
    );
    expect(output).toMatchSnapshot();
  });
  it('should call onChange when input is changed', () => {
    const onChange = jest.fn(value => expect(value).toBe(10))
    const output = renderer.create(
      <MoodNew
        onChange={onChange}
        onSubmit={() => void 0}
        value={20}
        lastRecorded={{ date: 1542226365908, value: 5 }}
      />
    );
    output.root.findByProps({ 'data-test': 'moodBar' }).props.onChange({ currentTarget: { value: 10 } })
    expect(onChange).toHaveBeenCalled()
  })
  it('should call onSubmit when input is changed', () => {
    const onSubmit = jest.fn(value => expect(value).toBe(20))
    const event = { preventDefault: jest.fn() }
    const output = renderer.create(
      <MoodNew
        onSubmit={onSubmit}
        onChange={() => void 0}
        value={20}
        lastRecorded={{ date: 1542226365908, value: 5 }}
      />
    );
    output.root.findByProps({ 'data-test': 'form' }).props.onSubmit(event)
    expect(onSubmit).toHaveBeenCalled()
    expect(event.preventDefault).toHaveBeenCalled()
  })
});
