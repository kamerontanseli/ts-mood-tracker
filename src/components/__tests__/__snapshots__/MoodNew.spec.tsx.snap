// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoodNew should render 1`] = `
<div>
  <p>
    Last recorded: 
    5
  </p>
  <form
    data-test="form"
    onSubmit={[Function]}
  >
    <input
      data-test="moodBar"
      max="10"
      min="1"
      name="mood"
      onChange={[Function]}
      required={true}
      type="range"
    />
    <input
      type="submit"
      value="Save"
    />
  </form>
</div>
`;
